ec/ecresource/deploymentresource/planmodifiers/topology_dedicated_masters.go

```diff
@@ -0,0 +1,213 @@
+		return
+	}
+
+	var planElasticsearch es.ElasticsearchTF
+	resp.Diagnostics.Append(req.Plan.GetAttribute(ctx, path.Root("elasticsearch"), &planElasticsearch)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	dedicatedMastersThreshold := getDedicatedMastersThreshold(template)
+	if dedicatedMastersThreshold == 0 {
+		// No automatic dedicated masters management
+		return
+	}
+
+	nodesInCluster := countNodesInCluster(ctx, planElasticsearch, template)
+
+	if nodesInCluster < dedicatedMastersThreshold {
+		// Disable master tier
+		resp.Plan.SetAttribute(ctx,
+			path.Root("elasticsearch").AtName("master"),
+			types.ObjectNull(es.ElasticsearchTopologyAttrs()),
+		)
+	} else {
+		// Enable master tier
+		instanceConfiguration := getInstanceConfiguration(template, "master")
+		if instanceConfiguration == nil {
+			tflog.Debug(ctx, "UpdateDedicatedMasterTier: Could not enable master tier, as it has no instance-config.")
+			return
+		}
+
+		if instanceConfiguration.DiscreteSizes == nil {
+			return
+		}
+
+		// Set zones
+		zones := instanceConfiguration.MaxZones
+		if zones == 0 {
+			return
+		}
+		resp.Plan.SetAttribute(ctx,
+			path.Root("elasticsearch").AtName("master").AtName("zone_count"),
+			zones,
+		)
+
+		// Set Size
+		defaultSize := util.MemoryToState(instanceConfiguration.DiscreteSizes.DefaultSize)
+		resp.Plan.SetAttribute(ctx,
+			path.Root("elasticsearch").AtName("master").AtName("size"),
+			defaultSize,
+		)
+	}
+}
+
+func getDedicatedMastersThreshold(template models.DeploymentTemplateInfoV2) int32 {
+	dedicatedMastersThreshold := int32(0)
+	if template.DeploymentTemplate != nil && template.DeploymentTemplate.Resources != nil {
+		for _, e := range template.DeploymentTemplate.Resources.Elasticsearch {
+			if e.Settings == nil {
+				continue
+			}
+			dedicatedMastersThreshold = e.Settings.DedicatedMastersThreshold
+			break
+		}
+	}
+	return dedicatedMastersThreshold
+}
+
+func countNodesInCluster(ctx context.Context, esPlan es.ElasticsearchTF, template models.DeploymentTemplateInfoV2) int32 {
+	nodesInDeployment := int32(0)
+	tiers := []string{"hot_content", "coordinating", "warm", "cold", "frozen"}
+	for _, tier := range tiers {
+		var rawTopology types.Object
+		switch tier {
+		case "hot_content":
+			rawTopology = esPlan.HotContentTier
+		case "coordinating":
+			rawTopology = esPlan.CoordinatingTier
+		case "warm":
+			rawTopology = esPlan.WarmTier
+		case "cold":
+			rawTopology = esPlan.ColdTier
+		case "frozen":
+			rawTopology = esPlan.FrozenTier
+		}
+
+		if rawTopology.IsNull() || rawTopology.IsUnknown() {
+			continue
+		}
+
+		var topology es.ElasticsearchTopologyTF
+		diags := rawTopology.As(ctx, &topology, objectAsOptions)
+		if diags.HasError() {
+			tflog.Debug(ctx, "countNodesInCluster: Failed to read topology.", map[string]interface{}{"error": diags.Errors()})
+			continue
+		}
+
+		size, err := deploymentsize.ParseGb(topology.Size.ValueString())
```

Comment:
`size` can be omitted in config (so the planned value is "unknown" and parsing fails). I guess we need to take `size` from template in that case.

Reply:
I now load size and zone_count from the template as a fallback.